oeps/oep-0036-microservice-routing-layer.rst

```diff
@@ -0,0 +1,89 @@
+===================================
+OEP-36: Microservice Routing Layer
+===================================
+
+.. list-table::
+
+  * - OEP
+    - :doc:`oep-0036-microservice-routing-layer`
+  * - Title
+    - Microservice Routing Layer
+  * - Last-Modified
+    - 2019-04-25
+  * - Author
+    - Fred Smith <derf@edx.org>
+  * - Arbiter
+    - David Joy <djoy@edx.org>
+  * - Status
+    - Draft
+  * - Type
+    - Architecture
+  * - Created
+    - 2019-04-25
+
+Abstract
+========
+
+Proposes a key openedx component that will connect users to applications using a layer 7 (application layer/URL Based) routing framework. 
+
+Motivation
+==========
+
+Microservices are currently implemented using layer 4 (DNS) based routing. Each microservice occupies a unique hostname, with dedicated ELB or CDN infrastructure to support routing user traffic to the service. This pattern is simple to monitor, deploy, and manage, but slow and expensive to build, and regularly causes problems with cross service user information sharing. In Addition, modifying or splitting a monolithic service into microservices results in confusing URL changes and costly migrations.
+
+Specification
+=============
+
+A layer 7 routing application will be chosen that meets the following requirements:
+
+* High performance connection management capabilities, including support for slow clients or backends, and wire-speed throughput.
+* SSL termination on the router.
+* Flexible configuration backends including container service discovery, flat files, and filesystem routing.
+* Path based routing, including path subsets and regular expression matching
+
+This layer 7 routing application will be deployed on a common hostname.
+
+Rationale
+=========
+
+A layer 7 routing layer would allow us to separate the user facing URL or API endpoint from the backend implementation. It would allow microservices to exist that only manage a small number of endpoints, without requring developers to build out load balancing or CDN infrastructure. 
+
+We are currently using a layer 7 routing framework for the edx.org marketing site, allowing them to replace drupal generated pages with gatsby generated pages quickly and seamlessly.
+
+This change will allow developers, openedx operators and resellers to deploy new openedx sites quicker and easier than they currently can.  The router will provide a map for people to understand which services they need to deploy.
+
+
+
+Implementation
+==============
+
+* to be determined
+
+Future
+======
+
+Possible additional functionality that may be implemented at the l7 router:
+
+* Authentication
+* Rate limiting
+* Web Application Firewall
+
+
+Backwards Compatibility
+=======================
+
+Services will begin to move API and user endpoints behind this layer 7 router after declaring their top level routing path. eg.
+
+* the LMS would occupy /courses
+* prospectus would occupy /course, /school, /micromasters
```

Comment:
Since we'll be starting out the gate with stuff that's not applicable to folks in the community, it would be good to map out how we'll cleanly separate out the pieces of it that are generically Open edX vs. the pieces that are site-specific customizations. Things like:

* Can a site cleanly upgrade to Koa/Lilac without having to change any site-specific config for their own marketing pieces?
* What conventions should we have so that we don't introduce namespace conflicts later on? Should site-specific stuff be put on its own domain? With a known prefix? With an org-specific prefix (e.g. `/edx`)?

Reply:
I would like this to be entirely applicable to the community out of the gate.  Perhaps we should segment edx.org things into their own bucket, but since the plan is for us to take over the marketing site, I would like to keep their URLS functional.